src/LinuxCaptureService/LinuxCaptureService.cpp

```diff
@@ -147,6 +149,90 @@ class ProducerEventProcessorHijackingFunctionEntryExitForLinuxTracing
 
 }  // namespace
 
+orbit_capture_service::CaptureService::StopCaptureReason
+LinuxCaptureService::WaitForStopCaptureRequestOrMemoryThresholdExceeded(
+    grpc::ServerReaderWriter<orbit_grpc_protos::CaptureResponse, orbit_grpc_protos::CaptureRequest>*
+        reader_writer) {
+  // wait_for_stop_capture_request_thread_ below outlives this method, hence the shared pointers.
+  auto stop_capture_mutex = std::make_shared<absl::Mutex>();
+  auto stop_capture = std::make_shared<bool>(false);
+  auto stop_capture_reason = std::make_shared<std::optional<StopCaptureReason>>();
+
+  static const std::optional<uint64_t> mem_total_bytes = ReadMemTotalInBytesFromProcMeminfo();
+  std::thread memory_watchdog_thread;
+  if (!mem_total_bytes.has_value()) {
+    ERROR("Could not start memory watchdog because total physical memory is unknown");
+  } else {
+    uint64_t watchdog_threshold_bytes = mem_total_bytes.value() / 2;
+    static constexpr absl::Duration kWatchdogPollInterval = absl::Seconds(1);
```

Comment:
Nit: As far as I can see the scope of this variable can be reduced by declaring it further down, right?